csrc/id_graphs.h

```diff
@@ -0,0 +1,492 @@
+
+  virtual ~IdGraphVisitor() = default;
+
+ private:
+  IdGraph& id_graph_;
+  const VectorOfUniqueEntries<IterDomain*> sub_selection_;
+};
+
+// Statement sorting based on IdGraphVisitor, see warnings to IdGraph Visitor.
+class IdGraphStmtSort : public IdGraphVisitor {
+ public:
+  IdGraphStmtSort(
+      IdGraph& id_graph,
+      const VectorOfUniqueEntries<IterDomain*> sub_selection = {})
+      : IdGraphVisitor(id_graph, sub_selection) {
+    IdGraphVisitor::traverse();
+  }
+
+  ExprGroups exprs() {
+    return sorted_exprs;
+  }
+
+  IdGroups ids() {
+    return sorted_ids;
+  }
+
+  ~IdGraphStmtSort() override = default;
+
+ protected:
+  using IdGraphVisitor::handle;
+  void handle(IdGroup id_group) override {
+    sorted_ids.pushBack(id_group);
+  }
+
+  void handle(ExprGroup expr_group) override {
+    sorted_exprs.pushBack(expr_group);
+  }
+
+  ExprGroups sorted_exprs;
+  IdGroups sorted_ids;
+};
+
+// TODO: Comment is stale, update.
+//
+// There's three modes of these iter domain mappings all uniquely important in
+// the lowering process.
+//
+// For EXACT/PERMISSIVE mode consider:
+//
+// consumer[i0, b1] = producer[i0]
+// consumer->merge(0) (consumer will now be [i0 * b1])
+// When producer is replayed as consumer (the direction we use for mapping)
+// with BestEffortReplay forward_bcast_mismatch = True the producer to
+// consumer map will have both a mapping of consumer(i0) to producer(i0) as
+// well as consumer(i0*b1) to producer(i0). This latter mapping is important
+// for loop nest mappings as the consumer will generate a loop based on i0*b1
+// and the producer may be computeAt inside this loop nest. However, for
+// indexing we do not want these two maps as producer may be indexed as i0*i1
+// depending on the loop nest structure and how it was built. Therefore we
+// really need to carry (at least) two sets of maps around for lowering.
+//
+// LOOP mode is important if we have something like:
+// consumer[i0o, threadIdx.x{i0i}] = producer[i0o, threadIdx.y{i0i}](computeAt
+// = 1) which can easily happen when using shared memory. We want to make sure
+// that the iteration domain used for loop construction (concreteId) has the
+// proper parallelization strategy. In parallel mode we do typical iteration
+// domain mapping, however we remove from it any iteration domains outside the
+// computeAt of producer when mapping. This guarentees we won't map
+// IterDomains that could have different parallelization strategies. We also
+// propagate the parallel strategy in parallel mode so all mapped IDs that
+// must have the same parallel type, do.
+//
+// IdMappingMode::LOOP
+//   Only maps leaf axes to left of compute at
+//   Forward broadcast axes in replay
+// IdMappingMode::PERMISSIVE
+//   Forward broadcast axes in replay
+//   Map all iteration domains
+//   Always contain root mappings (otherwise they could have been forwarded in
+//   broadcast)
+// IdMappingMode::EXACT
```

Comment:
This may be nonsensical; if so please ignore :-) On the theme of incrementally building, and just trying to simplify or understand, I am wondering if we could view the building as happening on the built graphs, i.e. including edges. Currently we don't make much use of relations between equivalence classes when building multiple equiv classes at once under these various notions. What if instead we built a low level "exact" graph whose nodes are classes under the strictest form of equivalence. We would add edges based on all of the stuff we need to track: resolving broadcasts, inlined above in some tensor, etc. This is a mapping from one graph (the Fusion) to a new graph (Exact nodes plus lots of edges). We'd then have another step that maps this exact graph to a permissive map or a loop map etc, and we only need to worry about the delta between each notion of equivalence and how it relates to the retained edge information.

Reply:
Would this require strict supersets of equivalency? Not sure I'm following.

Reply:
My understand is that `IdMappingMode::EXACT` is the most restrictive (i.e. no broadcasts are mapped to non-broadcasts ever). So this is the "finest" partition and any of the others can be obtained via unions of the classes in the exact map. The rules for which classes to perform additional merges on can be viewed as binary relations on either the original IterDomains or on their exact classes. For example, the AlmostExact partition is formed like this:
```c++
// IdMappingMode::AlmostExact
//   Forward through broadcast axes, but not through to a non-broadcast axis
//     i.e. id{b1*i0}, id{i0} are mapped
//          id{i1*i0}, id{i0} are not mapped (this part is the difference from
//          PERMISSIVE)
//   Forward through split one axes, i.e. id{ceilDiv(i0, 1)}, id{i0} are mapped
```
When we build the Exact partition, we can collect binary relations indicating these types of correspondences between IterDomains as lists of pairs of IterDomains like `std::vector<std::pair<IterDomain*, IterDomain*> > split_by_ones;`. Then we can quickly derive the AlmostExact partition from Exact by merging the subsets corresponding to the pairs. This can be done at the level of equivalence classes by using `std::vector<std::pair<EquivClass<IterDomain>, EquivClass<IterDomain>> split_by_ones;` instead. This is how I've been thinking of "IterDomain graphs" with classes as nodes and these types of edges. We'd have other types of edges like "resolves broadcast", "appears_left_of_compute_at", etc.

The current approach is achieving the same thing I'm sure. I'm just trying to understand how the different mapping modes relate to one enough in terms of what do I need to merge from exact in order to build each of the others (and is that always possible or is exact not the finest partition involved)?

Reply:
Yes, and IEL is built because we know the loop and exact relationships, and we have to intersect them. We could compress storage at some point.